src/smt/set_defaults.cpp

```diff
@@ -52,59 +52,82 @@ void setDefaults(LogicInfo& logic, bool isInternalSubsolver)
   // implied options
   if (options::debugCheckModels())
   {
-    Notice() << "SmtEngine: setting checkModel" << std::endl;
     options::checkModels.set(true);
   }
   if (options::checkModels() || options::dumpModels())
   {
-    Notice() << "SmtEngine: setting produceModels" << std::endl;
     options::produceModels.set(true);
   }
   if (options::checkModels())
   {
-    Notice() << "SmtEngine: setting produceAssignments" << std::endl;
     options::produceAssignments.set(true);
   }
+  // unsat cores and proofs shenanigans
   if (options::dumpUnsatCoresFull())
   {
-    Notice() << "SmtEngine: setting dumpUnsatCores" << std::endl;
     options::dumpUnsatCores.set(true);
   }
-  if ((options::unsatCores() && options::unsatCoresNew())
-      || (options::checkUnsatCores() && options::checkUnsatCoresNew()))
+  if (options::checkUnsatCores() || options::dumpUnsatCores()
+      || options::unsatAssumptions())
   {
-    AlwaysAssert(false) << "Can't have both unsat cores modes, pick one.\n";
+    options::unsatCores.set(true);
   }
-  if (options::checkUnsatCores())
+
+  if (options::unsatCores()
+      && options::unsatCoresMode() == options::UnsatCoresMode::OFF)
   {
-    options::unsatCores.set(true);
+    if (options::unsatCoresMode.wasSetByUser())
+    {
+      Notice()
+          << "Overriding OFF unsat-core mode since cores were requested..\n";
+    }
+    options::unsatCoresMode.set(options::UnsatCoresMode::OLD_PROOF);
   }
-  if (options::checkUnsatCoresNew())
+
+  if (options::checkProofs() || options::dumpProofs())
   {
-    options::unsatCoresNew.set(true);
+    options::produceProofs.set(true);
   }
-  if (options::dumpUnsatCores() || options::unsatAssumptions())
+
+  if (options::produceProofs()
+      && options::unsatCoresMode() != options::UnsatCoresMode::FULL_PROOF)
   {
-    if (!options::unsatCoresNew())
+    if (options::unsatCoresMode.wasSetByUser())
     {
-      Notice() << "SmtEngine: setting unsatCores" << std::endl;
-      options::unsatCores.set(true);
+      Notice() << "Forcing full-proof mode for unsat cores mode since proofs "
+                  "were requested.\n";
     }
+    options::unsatCoresMode.set(options::UnsatCoresMode::FULL_PROOF);
   }
-  if (options::unsatCoresNew()
-      && ((options::produceProofs() && options::produceProofs.wasSetByUser())
-          || (options::checkProofs() && options::checkProofs.wasSetByUser())
-          || (options::dumpProofs() && options::dumpProofs.wasSetByUser())))
+
+  // set proofs on if not yet set
+  if (options::unsatCores() && !options::produceProofs()
+      && options::unsatCoresMode() != options::UnsatCoresMode::OLD_PROOF)
```

Comment:
I think this is wrong for ASSUMPTIONS unsat cores.

Reply:
I don't think I see how. It'd have been a scenario in which one e.g. passes `--produce-unsat-cores --unsat-cores-mode=assumptions`, which would require us turning proofs and the unsat core mode would not be changed (since proofs were not on in the test in the code block above). The initialization of the solver would be done as expected, I believe.

Reply:
So you're saying that this wouldn't happen? Maybe add an assert for this case? E.g. that proofs are enabled when we have unsat cores ASSUMPTIONS.